infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/log/LogWatcher.java

```diff
@@ -0,0 +1,137 @@
+
+  private final String namespace;
+  private final KubernetesClient client;
+  private final PodLogHandler logHandler;
+
+  // set of current watchers. This is used so we're able to cut-off the watchers from outside.
+  private final Set<LogWatch> currentWatchers = new HashSet<>();
+
+  // set of currently observed containers so we're not try to follow same container multiple-times.
+  private final Set<String> watchingContainers = new HashSet<>();
+
+  public LogWatcher(
+      KubernetesClientFactory clientFactory,
+      String workspaceId,
+      String namespace,
+      PodLogHandler handler)
+      throws InfrastructureException {
+    this.logHandler = handler;
+    this.client = clientFactory.create(workspaceId);
+    this.namespace = namespace;
+  }
+
+  @Override
+  public void handle(PodEvent event) {
+    final String podName = event.getPodName();
+
+    if (logHandler.matchPod(podName)) {
+      final String containerName = event.getContainerName();
+      if (containerName != null && !watchingContainers.contains(containerName)) {
+        watchingContainers.add(containerName);
+        LOG.trace(
+            "adding [{}] to watching containers now watching [{}]",
+            containerName,
+            watchingContainers);
+        pool.submit(new ContainerLogWatch(podName, containerName));
+      }
+    }
+  }
+
+  @Override
+  public void close() {
+    new Thread(
+            () -> {
+              try {
+                LOG.debug("Waiting 5s before exit to get all the logs");
+                Thread.sleep(5000);
+              } catch (InterruptedException e) {
+                LOG.error("Interrupted waiting for the logs", e);
+              } finally {
+                currentWatchers.forEach(LogWatch::close);
+                currentWatchers.clear();
+              }
+            })
+        .start();
+  }
+
+  private class ContainerLogWatch implements Runnable {
+
+    private final String podName;
+    private final String containerName;
+
+    private ContainerLogWatch(String podName, String containerName) {
+      this.podName = podName;
+      this.containerName = containerName;
+    }
+
+    /**
+     * Do the best effort to get the logs from the container. The method tries N times to get the
+     * logs from the container. If response on log request from the k8s is 40x, it possibly means
+     * that container is not ready to get the logs and we'll try again.
+     */
+    @Override
+    public void run() {
+      int retries = 0;
+      while (retries < 10) {
+        LOG.debug("try watching the logs [{}]", retries);
+        try (LogWatch log =
+            client
+                .pods()
+                .inNamespace(namespace)
+                .withName(podName)
+                .inContainer(containerName)
+                .watchLog()) {
+          currentWatchers.add(log);
+          if (!logHandler.handle(log.getOutput(), containerName)) {
```

Comment:
Couldn't the error handling of `if line contains "code:40"` be done here instead of forcing each log handler impl to do that on its own?

Reply:
yes, I've moved whole reading of input stream here and PodLogHandler is now responsible just for handling single messages